Show correct liqudation price in ui

bonomat opened the pull request:
resolves #2754

bonomat committed:
Add simple tests

We add those tests to show that we can simplify the calculations of the bankruptcy price and still get the same values.

Signed-off-by: Philipp Hoenisch <philipp@coblox.tech>

bonomat committed:
fix (quanto): calculate bankruptcy price

By simplifying the formula of how we compute the bankruptcy price we can pre-compute all values prior sending them out to the frontend.

Signed-off-by: Philipp Hoenisch <philipp@coblox.tech>

luckysori commented on model/src/payout_curve/quanto.rs:
```diff
@@ -621,6 +621,7 @@ mod api_tests {
 #[cfg(test)]
 mod unit_tests {
     use super::*;
+    use crate::payout_curve::ETHUSD_MULTIPLIER;
```
🔧 I introduced an arbitrary multiplier just below. Shouldn't need to import the `ETHUSD` one.

luckysori replied:
I think it shouldn't matter what constant you use to check for regressions!

luckysori commented on model/src/payout_curve/quanto.rs:
```diff
@@ -231,45 +230,15 @@ impl Curve {
-            self.initial_price,
-            self.multiplier,
-        );
+        let bankruptcy_price = bankruptcy_price_long(self.initial_price, self.leverage_long);
```
🤔 I think this will lead to the liquidation price not changing as we roll over and accumulate fees. 

As I mentioned in private, I think this works for the offer because there are minimal (perhaps zero) fees at that stage and it's just an informative figure.

But as we roll over I think we need to take into account how the effective margins change. Otherwise the liquidation price will remain the same. Maybe that's what we want, but that's not how we implemented this at first, so we at least need to think about this change.

da-kami replied:
As far as I can remember this is an old problem that boils down to the margin changing because the interest (i.e. fees) is deducted from the funds locked up (which is a constraint that cannot easily be changed in our solution). The liquidation price calculated changing with the effective margin is not a feature we actively perused so far but I suppose it would be mathematically more correct. I wonder if this is something that a user would expect in behavior compared to other trading solutions out there.

luckysori replied:
My intuition would be that it's more correct. It would be interesting to consider how this behaviour compares to the inverse curve as well.

I don't know what's best, but if we decide to change it, could you document this change and the implications in the commit message, @bonomat?

bonomat replied:
I think it just clicked, let me see if I can explain it. 

The formula I used is correct but @luckysori is right, we will need to take the margin into account when doing rollover as the new bankruptcy price will differ.

The initial margin was calculated with 
`margin=(n_contracts * initial_price * Multiplier) / leverage`

but over time, the margin is getting reduced (or increased) through the fee account. 

What we have to do is calculate the _effective leverage_, i.e. the inverse of the above formula: 

`effective_leverage=(n_contracts * initial_price * Multiplier) / margin`

So basically, if we take my simplified formula of calculating the bankruptcy price of

`=EntryPrice - (EntryPrice * ((1/leverage) + InitialMargin))` 

we would need to swap `leverage` with the `effective leverage` calculated using the reduced (or increased) margin when rolling over and should get an accurate bankruptcy price when rolling over.

Does this make sense? I've played around here: 

https://docs.google.com/spreadsheets/d/1Hm9BKz1EKymGpNyto0SNShQ-CdZPC9EJqWwklvEtQ-w/edit#gid=0


I think that's actually what @luckysori was doing with his formulas but I did not get it.

bonomat replied:
The same applies for inverse contracts as well.

bonomat replied:
Implemented in [`f2599c1` (#2772)](https://github.com/itchysats/itchysats/pull/2772/commits/f2599c163bf6761aeca49e84953692625ffe7575)

da-kami commented on model/src/cfd.rs:
```diff
@@ -1965,18 +1953,8 @@ pub fn calculate_short_liquidation_price(
-                initial_price,
-                multiplier,
-            );
+            let liquidation_price = quanto::bankruptcy_price_short(initial_price, leverage);
```
🙃 nit: maybe we can harmonize the names and order of parameters of these functions, given that `inverse::calculate_short_liquidation_price(leverage, initial_price)` is named different and the parameters are passed flipped.

bonomat committed:
feat/fix (quanto): Use effective leverage

For calculating the liquidation price we have to take the updated fee margin (reduced or increased due to fee account) into account. This means, the liquidation price will change in the lifetime of a position, i.e. on every rollover.

Note: this commit also adds . A proptest which was failing before but is fixed with this commit.

luckysori commented on model/src/cfd.rs:
```diff
@@ -1927,7 +1927,13 @@ pub fn calculate_long_liquidation_price(
             let initial_price = initial_price.to_u64();
 
-            let liquidation_price = quanto::bankruptcy_price_long(initial_price, leverage);
+            // TODO: this liquidation price is potentially not in-line with the _actual_
+            // liquidation price as computed in the payout curve. The problem is we do not have the
+            // fee account available at this place and always compute the liquidation
+            // price based on the initial margin.
+            // Same for short liquidation price a few lines below.
```
🔧 I would move this to the docstring and not mark it as a TODO.

luckysori commented on model/src/payout_curve/quanto.rs:
```diff
@@ -369,8 +407,12 @@ pub fn bankruptcy_price_short(initial_price: u64, leverage: Leverage) -> u64 {
-    initial_price / leverage.0 as u64
+/// Note: the resulting bankruptcy price is rounded to 0 decimal points.
+fn bankruptcy_price_shift(initial_price: u64, leverage: Decimal) -> u64 {
+    (Decimal::from(initial_price) / leverage)
+        .round_dp_with_strategy(0, RoundingStrategy::ToZero)
+        .to_u64()
+        .expect("Bankruptcy price to fit into u64")
```
🙃 I think it's clearer if we convert everything to `Decimal` first, then have the formula in a separate line, and finally convert the result into the return type. That's what I've tried to do in this file.

luckysori commented on model/src/payout_curve/quanto.rs:
```diff
@@ -369,8 +407,12 @@ pub fn bankruptcy_price_short(initial_price: u64, leverage: Leverage) -> u64 {
 /// depend on the party's position.
-fn bankruptcy_price_shift(initial_price: u64, leverage: Leverage) -> u64 {
-    initial_price / leverage.0 as u64
+/// Note: the resulting bankruptcy price is rounded to 0 decimal points.
```
🔧 I think this line is unnecessary because the return type doesn't support decimal points.

luckysori commented on model/src/payout_curve/quanto.rs:
```diff
@@ -230,19 +231,56 @@ impl Curve {
+    /// formula to compute the initial margin in the first place: `margin=(n_contracts *
+    /// initial_price * multiplier) / leverage` --> `effective_leverage=(n_contracts *
+    /// initial_price * multiplier) / margin`
+    fn effective_leverage(&self, margin: Amount) -> Result<Decimal> {
+        let effective_leverage =
+            (Decimal::from(self.n_contracts) * Decimal::from(self.initial_price) * self.multiplier)
+                / Decimal::from_f64(margin.as_btc())
+                    .context("Could not convert initial margin to decimal")?;
+        Ok(effective_leverage)
+    }
```
🔧 This is just a generic leverage formula. It just so happens that you pass in the effective margin wherever you call it, but that's circumstantial.

I think we should have this method should therefore be a private free function.

luckysori commented on model/src/payout_curve/quanto.rs:
```diff
@@ -230,19 +231,56 @@ impl Curve {
+    /// initial_price * multiplier) / leverage` --> `effective_leverage=(n_contracts *
+    /// initial_price * multiplier) / margin`
+    fn effective_leverage(&self, margin: Amount) -> Result<Decimal> {
+        let effective_leverage =
+            (Decimal::from(self.n_contracts) * Decimal::from(self.initial_price) * self.multiplier)
+                / Decimal::from_f64(margin.as_btc())
+                    .context("Could not convert initial margin to decimal")?;
+        Ok(effective_leverage)
```
🙃 If you put the formula on a separate line from all the conversions you probably don't need to add it to the documentation.

luckysori reviewed (approved):
It looks like it works, but I think we can refactor things slightly.
